package.json

```diff
@@ -15,6 +15,7 @@
     "front-matter": "^4.0.2",
     "gatsby": "^2.24.47",
     "gatsby-image": "^2.4.16",
+    "gatsby-plugin-layout": "^1.3.12",
```

Comment:
So i looked a little bit into the documentation, and it says: 
> This plugin enables adding components which live above page components and persist across page changes.

And I was wondering if you would be able to explain this more to me?

Reply:
Yes, [this section](https://www.gatsbyjs.com/plugins/gatsby-plugin-layout/?=layout#why-would-you-want-to-reimplement-the-v1-layout-behavior) of those docs help explain it a bit more in-depth, but lets illustrate with a couple examples to try and explain why this plugin helps.

Lets take the following React code:
```js
const Routes = ({ children }) => {
  const [route, setRoute] = useState('/')

  return route === '/' ? (
    <Home />
  ) : page === '/one' ? (
    <PageOne />
  ) : (
    <NotFound />
  )
}

const Home = () => {
  return (
    <>
     <Header />
     My content
    </>
  )
}

const PageOne = () => {
  return (
    <>
     <Header />
     Page 1
    </>
  )
}

const NotFound = () => {
  return (
    <>
     <Header />
     404
    </>
  )
}
```

Note here how each of the "route" components render the `<Header />` inside of their render function. Lets say the state changes from `/` to `/one`. Under the hood, React is going to fully unmount the `<Home />` component, then mount the `<PageOne />` component. Because it is fully unmounting and remounting a different component, it means that the `<Header />` inside `<Home />` is also fully unmounted, then remounted with `<PageOne />`. This presents a problem if, say, `Header` had some local state inside of it. The state gets lost during the transition between the 2 components.

So how might you fix this? The answer is to pull the `Header` up to the `Routes` component instead so that the `<Header />` component doesn't unmount between route changes:

```js
const Routes = ({ children }) => {
  const [route, setRoute] = useState('/')

  return (
    <>
      <Header />
      {route === '/' ? (
        <Home />
      ) : page === '/one' ? (
        <PageOne />
      ) : (
        <NotFound />
      )
    }
  )
}

const Home = () => {
  return 'My content'
}

const PageOne = () => {
  return 'Page 1'
}

const NotFound = () => {
  return '404'
}
```

Note here how the `Header` is no longer being rendered by the individual routes, but rather the container of these routes.

This is essentially what this plugin is doing. Gatsby manages the top-level "route" components for us and gives us the ability to specify what goes on a page. Out-of-the-box rendering pages look a little bit like the first example: If we render a `Layout` on an individual page, it means that transitioning to a new route will unmount the `Layout` and remount on the new path. This is not ideal of you want to maintain state between route changes. This plugin essentially "fixes" this and allows you to specify the layout higher up in the component hierarchy so that route changes don't force an unmount/remount of the layout.

Does that help?

Reply:
Yes that explains a lot! Thank you.